### PR TITLE
New challenges chapters 7-13

### PR DESCRIPTION
**Pull request by shadowc**
Turbo Challenges for chapters 7 to 13

### CODE REVIEW

File: challenges-7-plus.yml [hunk 1]
```diff
@@ -0,0 +1,277 @@
+preview:
+  question: |
+    Hey! I've prepared this template that needs NO preview from Turbo when it loads:
+    
+    ```twig
+    {% extends 'base.html.twig' %}
+    
+    {% block title%}Play Pong in the Browser for Free!{% endblock %}
+    
+    {% block metas %}
```
Comment (weaverryan) [inline challenges-7-plus.yml:10]:
```suggestion
    {% block metas %}{% endblock %}
```

File: challenges-7-plus.yml [hunk 2]
```diff
@@ -0,0 +1,277 @@
+preview:
+  question: |
+    Hey! I've prepared this template that needs NO preview from Turbo when it loads:
```
Comment (weaverryan) [inline challenges-7-plus.yml:3]:
```suggestion
    Hey! I've prepared this template for a shiny new Pong page!
```

File: challenges-7-plus.yml [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    {% block metas %}
+    
+    {% endblock %}
+    
+    {% block body %}
+    <div id="pong_game_app"></div>
+    {% endblock %}
+    ```
+    
+    I even added a `block metas` so that I can add the proper `head` tag to avoid this... but,
```
Comment (weaverryan) [inline challenges-7-plus.yml:19]: But now... I want to avoid showing a "preview" if the user navigates back to this page. I've added
a `block metas` so that I an add the proper tag to the `head` element to avoid the preview... but

File: challenges-7-plus.yml [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    I can't seem to remember exactly how this is!
```
Comment (weaverryan) [inline challenges-7-plus.yml:20]:
```suggestion
    I can't seem to remember exactly what that tag should look like!
```

File: challenges-7-plus.yml [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    <meta name="turbo-visit-control" content="reload">
+    ```
+    
+    To enforce a complete reload every time turbo visits this page. Check the Turbo Drive reference
+    for more information on these!
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          ```twig
```
Comment (weaverryan) [inline challenges-7-plus.yml:39]: We can use `html` for all of these. I think that might highlight better (even though, in theory, twig should highlight the same)

File: challenges-7-plus.yml [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    {% endblock %}
+    ```
+    
+    I even added a `block metas` so that I can add the proper `head` tag to avoid this... but,
+    I can't seem to remember exactly how this is!
+    
+    What is the actual `meta` tag that we need to add to stop turbo previews?
+  explanation: |
+    You can also set this meta to contain the value `no-cache` which means that it won't generate
+    any previews at all (not even for the back button!)
```
Comment (weaverryan) [inline challenges-7-plus.yml:25]:
```suggestion
    any snapshot cache at all (not even for the back button!)
```

File: challenges-7-plus.yml [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+    Other metas that we can leverage with turbo are:
+    
+    ```twig
+    <meta name="turbo-visit-control" content="reload">
+    ```
+    
+    To enforce a complete reload every time turbo visits this page. Check the Turbo Drive reference
```
Comment (weaverryan) [inline challenges-7-plus.yml:33]:
```suggestion
    This enforces a full page reload every time turbo visits this page. Check the Turbo Drive reference
```

File: challenges-7-plus.yml [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    working.
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          At the moment right before leaving the page, Turbo clones the DOM using `cloneNode`,
+          which does *not* include event listeners added from JavaScript.
+        b: |
+          At page load, Turbo clones the DOM using a "deep clone" algorithm, but
+          since JavaScring hasn't executed yet, we don't get any listeners adde dy it.
```
Comment (weaverryan) [inline challenges-7-plus.yml:74]:
```suggestion
          since JavaScript hasn't executed yet, no listeners are included in the preview.
```

File: challenges-7-plus.yml [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+        c: |
+          At the moment right before leaving the page, Turbo clones the DOm using a "deep
```
Comment (weaverryan) [inline challenges-7-plus.yml:76]:
```suggestion
          At the moment right before leaving the page, Turbo clones the DOM using a "deep
```

File: challenges-7-plus.yml [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+          clone" algorithm, but removes any event listeners added by JavaScript because
+          it relies on Stimulus to add them back.
+        d: |
+          At page load, and after all JavaScript has been parsed and executed, Turbo clones
+          the DOM using a "deep clone" algorithm, but it removes any event listeners added by
```
Comment (weaverryan) [inline challenges-7-plus.yml:81]:
```suggestion
          the DOM using a "deep clone" algorithm. It then removes any event listeners added by
```

File: challenges-7-plus.yml [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+          JavaScript because ir relies on Stimulus to add them back.
```
Comment (weaverryan) [inline challenges-7-plus.yml:82]:
```suggestion
          JavaScript because it relies on Stimulus to add them back.
```

File: challenges-7-plus.yml [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+          ```twig
+          <meta name="cache" content="no-preview">
+          ```
+      correct: c
+
+modal_preview:
+  question: |
+    What happens when Turbo creates a preview of a page?
+  explanation: |
+    This is important because you need to make sure your page functionality is preserve
```
Comment (weaverryan) [inline challenges-7-plus.yml:60]:
```suggestion
    This is important because you need to make sure your JavaScript continues to work when
    a snapshot is restored (especially when hitting the back and forward buttons). Because `cloneNode()`
    is used, any event listeners (like a click listener) on the original page will *not* be included on
    the "restored" page. This means your JavaScript needs to be smart enough to continue working
    even in this situation. 
```

File: challenges-7-plus.yml [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+      correct: a
+  
+modal_preview_2:
+  question: |
+    In my web page, I decided to create my own modal. It's just a `div` element
+    that I set to `display: block` whenever I want to show it. Not a big deal!
+    
+    ```twig
+    <div id="my_great_modal" style="display: none">
+        <h1>Welcome our Pizza Section!</h1>
```
Comment (weaverryan) [inline challenges-7-plus.yml:92]:
```suggestion
        <h1>Welcome to our Pizza Section!</h1>
```

File: challenges-7-plus.yml [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+        </button>
+    </div>
+    ```
+    
+    I then use JavaScript to discretely show this popup under certain circumstances (i.e. when
+    I detect that a customer might not love pizza!).
+    
+    But, when the user leaves the page and then goes back, Turbo is showing this modal
+    *even* when the user might have proven otherwise that he is indeed a Pizzal lover!
+    I don't want that!
```
Comment (weaverryan) [inline challenges-7-plus.yml:110]: I really like this example! But this paragraph confused me - it may just be too clever. The problem is simply that if the modal is open, and I click to another page, then click back to the page, the modal would be shown for a moment (the preview) before it disappeared, right?

File: challenges-7-plus.yml [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+    
+    How can I hide this modal every time Turbo shows my page preview?
+  explanation: |
+    Leveraging the turbo event `turbo:before-cache` is the best solution for things like
+    these, because you can clean up and prepare your snapshow before turbo takes it so that
```
Comment (weaverryan) [inline challenges-7-plus.yml:115]:
```suggestion
    these, because you can clean up and prepare your snapshot before turbo takes it so that
```

File: challenges-7-plus.yml [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+    you ensure that you will always see the right state!
+    
+    Other seemingly useful events, like `turbo:load` might look useful, but they will
```
Comment (weaverryan) [inline challenges-7-plus.yml:118]:
```suggestion
    Other seemingly events, like `turbo:load` might look useful, but they will
```

File: challenges-7-plus.yml [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+          document.addEventListener('turbo:load', (event) => {
+              document.getElementById('my_great_modal').display = 'none';
+          });
+          ```
+      correct: b
+
+cleaned_preview:
+  question: |
+    In our video, we show how closing a bootstrap dialog before exiting a page,
+    so that we can have a cleaned preview, doesn't quite work. When we came
```
Comment (weaverryan) [inline challenges-7-plus.yml:153]:
```suggestion
    (so that we can have a cleaned preview) doesn't quite work. When we came
```

This just helped me read it.

File: challenges-7-plus.yml [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+    back to the page, we saw a greyed invalidation div covering it!
```
Comment (weaverryan) [inline challenges-7-plus.yml:154]:
```suggestion
    back to the page, we saw a greyed overlay div covering it!
```

You're talking about the semi-transparent `overlay`, right?

File: challenges-7-plus.yml [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+    
+    Fixing this required some more hacking on our part in order to have it working
```
Comment (weaverryan) [inline challenges-7-plus.yml:156]:
```suggestion
    Fixing this required some more hacking to get it working
```

File: challenges-7-plus.yml [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+    properly.
+    
+    Why did we get this weird behavior?
+  explanation: |
+    The `turbo:before-cache` event is synchronous, which means that what you do in that
+    event handler will happen instantly. If in that event you call some animation
```
Comment (weaverryan) [inline challenges-7-plus.yml:162]:
```suggestion
    event handler will happen instantly. If, in that event, you call some animation
```

File: challenges-7-plus.yml [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+    (like the bootstrap `close` function), then the animation won't finish executing,
+    preventing any callbacks after the animation from cleaning up the page.
```
Comment (weaverryan) [inline challenges-7-plus.yml:164]: This might require a bit more explanation, since it's so subtle.

> , then the animation won't finish executing. This can cause two potential problems. First,
> the snapshot will be taken before the element that is animating (e.g. the modal closing) 
> is finished animating. And second, as in this case, Bootstrap is "listening" for the animation
> to finish so that it can do other clean up, like removing the overlay. But because the element
> was removed from the page, the animation never "finishes" and the cleanup never happens.
> Weird, eh?

File: challenges-7-plus.yml [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+    
+    It's important to know this for any third party code that we use, so that when
+    preparing a snapshop on the `before-cache` event goes smoothly!
```
Comment (weaverryan) [inline challenges-7-plus.yml:167]: I think we can remove this

File: challenges-7-plus.yml [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+    
+    > Alternatively, you can disable cache on certain pages if the clean up
+    > process seems too complicated
```
Comment (weaverryan) [inline challenges-7-plus.yml:170]: Btw, there is also a new feature maybe we could mention here - this is in beta 8 https://github.com/hotwired/turbo/pull/238

File: challenges-7-plus.yml [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          This happened because bootstrap has a bug where some times calling `close`
```
Comment (weaverryan) [inline challenges-7-plus.yml:175]:
```suggestion
          This happened because bootstrap has a bug where calling `close`
```

File: challenges-7-plus.yml [hunk 25]
```diff
@@ -0,0 +1,277 @@
...
+          on a modal doesn't close it properly.
```
Comment (weaverryan) [inline challenges-7-plus.yml:176]:
```suggestion
          on a modal sometimes doesn't close it properly.
```

File: challenges-7-plus.yml [hunk 26]
```diff
@@ -0,0 +1,277 @@
...
+        b: |
+          This happened because Turbo has a bug where some times our code inside a
```
Comment (weaverryan) [inline challenges-7-plus.yml:178]:
```suggestion
          This happened because Turbo has a bug where sometimes our code inside a
```

File: challenges-7-plus.yml [hunk 27]
```diff
@@ -0,0 +1,277 @@
...
+          `turbo:before-cache` event doesn't execute properly.
```
Comment (weaverryan) [inline challenges-7-plus.yml:179]:
```suggestion
          `turbo:before-cache` event doesn't execute quickly enough.
```

File: challenges-7-plus.yml [hunk 28]
```diff
@@ -0,0 +1,277 @@
...
+        c: |
+          This happenbed because we called `close` on our dialog and forgot to make the
```
Comment (weaverryan) [inline challenges-7-plus.yml:181]:
```suggestion
          This happened because we called `close` on our dialog and forgot to make the
```

File: challenges-7-plus.yml [hunk 29]
```diff
@@ -0,0 +1,277 @@
...
+          `turbo:before-cache` callback asynchronous.
+        d: |
+          This happened because we called `close` on our dialog right inside
+          `turbo:before-cache` and the closing animation didn't have time to
+          finish before the cache was saved.
+      correct: d
+
+js_widget:
+  question: |
+    In this video, we've added a Weather Widget sanippet into our page and saw how it
```
Comment (weaverryan) [inline challenges-7-plus.yml:191]:
```suggestion
    In this video, we've added a weather widget snippet to our page and saw how it
```

File: challenges-7-plus.yml [hunk 30]
```diff
@@ -0,0 +1,277 @@
...
+    only work the first time our page loaded. When we clicked on subsecuent pages, it
```
Comment (weaverryan) [inline challenges-7-plus.yml:192]:
```suggestion
    only works the first time our page loaded. When we clicked on subsequent pages, it
```

File: challenges-7-plus.yml [hunk 31]
```diff
@@ -0,0 +1,277 @@
...
+    stopped working!
+    
+    Why is this so?
```
Comment (weaverryan) [inline challenges-7-plus.yml:195]:
```suggestion
    Why?
```

File: challenges-7-plus.yml [hunk 32]
```diff
@@ -0,0 +1,277 @@
...
+    
+    > When we forced our snippet to add the `script` tag without any check, it effectively
+    > added a new copy of itself into our `head` *every* single time we change the page.
+    > This worked, but now we have a new problem: Our `head` is growing forever!
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          Because the `script` tag added by our snippet of code is only added once, hence
+          it only executes once.
```
Comment (weaverryan) [inline challenges-7-plus.yml:211]:
```suggestion
          it only executes once and never re-initializes the widget on subsequent pages.
```

File: challenges-7-plus.yml [hunk 33]
```diff
@@ -0,0 +1,277 @@
...
+        b: |
+          Because the snippet of code added by us is only executed once, hence it only
+          adds the `script` tag once
```
Comment (weaverryan) [inline challenges-7-plus.yml:214]: Hmm, I'm having a hard time distinguishing between A and B.

> `script` tag added by our snippet
vs
> snippet of code adde by us

File: challenges-7-plus.yml [hunk 34]
```diff
@@ -0,0 +1,277 @@
...
+        c: |
+          Because Turbo recognizes similar scripts between page views and avoids re-executing
+          them.
+        d: |
+          The widget snipped worked as expected. It did not stop working after subsecuent
```
Comment (weaverryan) [inline challenges-7-plus.yml:219]:
```suggestion
          The widget snipped worked as expected. It did not stop working after subsequent
```

File: challenges-7-plus.yml [hunk 35]
```diff
@@ -0,0 +1,277 @@
...
+          page views.
```
Comment (weaverryan) [inline challenges-7-plus.yml:220]:
```suggestion
          page views: we even verified that the `a` tag was present!
```

File: challenges-7-plus.yml [hunk 36]
```diff
@@ -0,0 +1,277 @@
...
+
+js_widget:
+  question: |
+    In this video, we've added a Weather Widget sanippet into our page and saw how it
+    only work the first time our page loaded. When we clicked on subsecuent pages, it
+    stopped working!
+    
+    Why is this so?
+  explanation: |
+    In our video, the snipped of code is executed every time we load a page because it's
```
Comment (weaverryan) [inline challenges-7-plus.yml:197]:
```suggestion
    In our video, the snippet of code is executed every time we load a page because it's
```

File: challenges-7-plus.yml [hunk 37]
```diff
@@ -0,0 +1,277 @@
...
+    in our `body` element! But in it there's a check before adding the `script` to our
```
Comment (weaverryan) [inline challenges-7-plus.yml:198]:
```suggestion
    in our `body` element! But inside the snipped itself, there's a check before adding the `script` to our
```

File: challenges-7-plus.yml [hunk 38]
```diff
@@ -0,0 +1,277 @@
...
+    `head`. This `script` is only added if it's not already present. This is why the
```
Comment (weaverryan) [inline challenges-7-plus.yml:199]:
```suggestion
    `head`: the `script` is only added if it's not already present. This is why the
```

File: challenges-7-plus.yml [hunk 39]
```diff
@@ -0,0 +1,277 @@
...
+    `script` tag was only added once, and therefore, only executed once between our page
```
Comment (weaverryan) [inline challenges-7-plus.yml:200]:
```suggestion
    `script` tag was only added once, and therefore, only executed one time no matter
    how many pages we visit with Turbo.
```

File: challenges-7-plus.yml [hunk 40]
```diff
@@ -0,0 +1,277 @@
...
+    views!
+    
+    > When we forced our snippet to add the `script` tag without any check, it effectively
+    > added a new copy of itself into our `head` *every* single time we change the page.
+    > This worked, but now we have a new problem: Our `head` is growing forever!
```
Comment (weaverryan) [inline challenges-7-plus.yml:205]:
```suggestion
    > This worked, but now we have a new problem: our `head` is growing forever!
```

File: challenges-7-plus.yml [hunk 41]
```diff
@@ -0,0 +1,277 @@
...
+          Because Turbo recognizes similar scripts between page views and avoids re-executing
+          them.
+        d: |
+          The widget snipped worked as expected. It did not stop working after subsecuent
+          page views.
+      correct: a
+
+analytics:
+  question: |
+    When does the event `turbo:before-render` fire?
```
Comment (weaverryan) [inline challenges-7-plus.yml:225]:
```suggestion
    When does the `turbo:before-render` event fire?
```

File: challenges-7-plus.yml [hunk 42]
```diff
@@ -0,0 +1,277 @@
...
+    multiple_choice:
+      choices:
+        a: |
+          It fires every time turbo has to render a page, but before its new contents have
+          arrived via Ajax.
+        b: |
+          It fires once per session, before Turbo initializes its JavaScript engine.
+        c: |
+          It fires every time that turbo has received a new page via Ajax and before it
+          replaces the `body` element with the new `body` contents.
```
Comment (weaverryan) [inline challenges-7-plus.yml:245]: I can't remember: does it fire on back/forward button? What about when a preview is temporarily shown? I might be interesting to mention these - even potentially make them part of the question/answer if it feels important.

File: challenges-7-plus.yml [hunk 43]
```diff
@@ -0,0 +1,277 @@
...
+        d: |
+          The widget snipped worked as expected. It did not stop working after subsecuent
+          page views.
+      correct: a
+
+analytics:
+  question: |
+    When does the event `turbo:before-render` fire?
+  explanation: |
+    This is very usefull because you can access the contents of the new `body` element
```
Comment (weaverryan) [inline challenges-7-plus.yml:227]:
```suggestion
    This is useful because you can access the contents of the new `body` element
```

File: challenges-7-plus.yml [hunk 44]
```diff
@@ -0,0 +1,277 @@
...
+    that is about to replace the old one, so that you can tweak it and modify it
+    before any JavaScript is executed!
+    
+    > You can access the "about to be added" `body` elements with `event.detail.newBody`!
```
Comment (weaverryan) [inline challenges-7-plus.yml:231]:
```suggestion
    > You can access the "about to be added" `body` element with `event.detail.newBody`!
```

File: challenges-7-plus.yml [hunk 45]
```diff
@@ -0,0 +1,277 @@
...
+          It fires every time that turbo has received a new page via Ajax, after it
+          replaces the `body` element but *before* executing any JavaScript in it
+      correct: c
+
+analytics_2:
+  question: |
+    At the end of our video, we talk about analytics and show an example of how you can
+    execute code *once* (and only once) on every page load (including the first page load!)
+    
+    What is this event and why others don't work?
```
Comment (weaverryan) [inline challenges-7-plus.yml:256]:
```suggestion
    What is this event? And what is the *best* explanation for why other events won't work?
```

File: challenges-7-plus.yml [hunk 46]
```diff
@@ -0,0 +1,277 @@
...
+  explanation: |
+    `turbo:laod` is very usefull if we need to execute some code on a "per view" basis. It will
+    execute on every page view and will *not* repeat itself if a page has a preview. *Just* what
+    we need to get our analytics code working!
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          The event to use is `turbo:load`. `turbo:before-render` and `turbo:render` do NOT execute
+          on the first load. The *also* execute twice if page has a preview to show first!
```
Comment (weaverryan) [inline challenges-7-plus.yml:266]:
```suggestion
          on the first load. They *also* execute twice if a page has a preview to show first!
```

**Review by weaverryan — COMMENTED** [review]
cool 👍

File: challenges-7-plus.yml (hunk 16 above)
Comment (shadowc) [reply challenges-7-plus.yml:121]: Hey! I don't understand "Other seemingly events" looks like this would be missing an adverb?

Change for "Other events, like `turbo:load` might look useful, but they will"

File: challenges-7-plus.yml (hunk 18 above)
Comment (shadowc) [reply challenges-7-plus.yml]: haha, yes!

File: challenges-7-plus.yml (hunk 37 above)
Comment (shadowc) [reply challenges-7-plus.yml:201]: "snippet" :D

File: challenges-7-plus.yml (hunk 14 above)
Comment (shadowc) [reply challenges-7-plus.yml:113]: Indeed you are right. Let me rephrase it.

File: challenges-7-plus.yml (hunk 33 above)
Comment (shadowc) [reply challenges-7-plus.yml:218]: Rephrasing it to:

```
        a: |
          Because the `<script>` tag added by our snippet of code is only added once, hence
          it only executes once.
        b: |
          Because the snippet of code added by us to the `<bode>` element is only executed once,
          hence it only adds the `<script>` tag to our `<head>` once.
```

File: challenges-7-plus.yml (hunk 16 above)
Comment (weaverryan) [reply challenges-7-plus.yml]: Yes, you're correct - seemingly should now be gone :)

File: challenges-7-plus.yml (hunk 42 above)
Comment (shadowc) [reply challenges-7-plus.yml]: I rephrased it based on the information you provide on the video!
